clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp

```diff
@@ -2060,13 +2103,38 @@ Expected<SmallVector<StringRef>> linkAndWrapDeviceFiles(
         return SplitModulesOrErr.takeError();
 
       auto &SplitModules = *SplitModulesOrErr;
-      for (size_t I = 0, E = SplitModules.size(); I != E; ++I) {
+      const llvm::Triple Triple(LinkerArgs.getLastArgValue(OPT_triple_EQ));
+      for (unsigned I = 0; I < SplitModules.size(); ++I) {
         SmallVector<StringRef> Files = {SplitModules[I].ModuleFilePath};
-        auto LinkedFileFinalOrErr =
+        StringRef Arch = LinkerArgs.getLastArgValue(OPT_arch_EQ);
+        if (Arch.empty())
+          Arch = "native";
+        SmallVector<std::pair<StringRef, StringRef>, 4> BundlerInputFiles;
+        auto ClangOutputOrErr =
             linkDevice(Files, LinkerArgs, true /* IsSYCLKind */);
-        if (!LinkedFileFinalOrErr)
-          return LinkedFileFinalOrErr.takeError();
-        SplitModules[I].ModuleFilePath = *LinkedFileFinalOrErr;
+        if (!ClangOutputOrErr)
+          return ClangOutputOrErr.takeError();
+        if (Triple.isNVPTX()) {
+          auto VirtualArch =
```

Comment:
Can we move the NVPTX code inside the if into a separate function?

Reply:
Is there a specific reason for this change?

Thanks

Reply:
I thought for the loop was getting a bit complicated

Reply:
I revisited the code. I agree that the loop is getting a bit complicated. But I am not able to come up with a 'logical' split to move the NVPTX case into a new function.